lib/workos/webhooks.rb

```diff
@@ -0,0 +1,148 @@
+      #
+      # @return [WorkOS::Webhook]
+      # rubocop:enable Layout/LineLength
+      sig do
+        params(
+          payload: String,
+          sig_header: String,
+          secret: String,
+          tolerance: Integer,
+        ).returns(WorkOS::Webhook)
+      end
+      def construct_event(
+        payload:,
+        sig_header:,
+        secret:,
+        tolerance: DEFAULT_TOLERANCE
+      )
+        verify_header(payload: payload, sig_header: sig_header, secret: secret, tolerance: tolerance)
+        WorkOS::Webhook.new(payload)
+      end
+
+      private
+
+      sig do
+        params(
+          payload: String,
+          sig_header: String,
+          secret: String,
+          tolerance: Integer,
+        ).returns(T::Boolean)
+      end
+      # rubocop:disable Metrics/MethodLength
+      def verify_header(
+        payload:,
+        sig_header:,
+        secret:,
+        tolerance: DEFAULT_TOLERANCE
+      )
+        begin
+          timestamp, signature_hash = get_timestamp_and_signature_hash(sig_header: sig_header)
+        rescue StandardError
+          raise WorkOS::SignatureVerificationError.new(
+            message: 'Unable to extract timestamp and signature hash from header',
+          )
+        end
+
+        if signature_hash.empty?
+          raise WorkOS::SignatureVerificationError.new(
+            message: 'No signature hash found with expected scheme v1',
+          )
+        end
+
+        if timestamp < Time.now - tolerance
+          raise WorkOS::SignatureVerificationError.new(
+            message: 'Timestamp outside the tolerance zone',
+          )
+        end
+
+        expected_sig = compute_signature(timestamp: timestamp, payload: payload, secret: secret)
+        unless expected_sig.eql?(signature_hash)
```

Comment:
I think we want to use a constant-time comparison here, and I don't think `String#eql?` is one.

I know Rails provides [`ActiveSupport::SecurityUtils.secure_compare`](https://api.rubyonrails.org/classes/ActiveSupport/SecurityUtils.html#method-c-secure_compare), but maybe there's an equivalent in the Ruby core library or in the deps we already have.

Reply:
Did a little research, and I think we can use [`OpenSSL.secure_compare`](https://ruby-doc.org/stdlib-2.4.0/libdoc/openssl/rdoc/OpenSSL.html#method-c-secure_compare) here instead of `String.eql?`.

Another alternative is to compare instances of `OpenSSL::HMAC`, [as it implements `==` using a constant time comparison](https://ruby-doc.org/stdlib-2.4.0/libdoc/openssl/rdoc/OpenSSL/HMAC.html#3D-3D-method). Either one should do the trick, but maybe the first one is easier and more explicit.

Reply:
I'm using `OpenSSL.secure_compare` now but getting a Sorbet error. Investigating now, but do you have any insight?

Change:

```
unless OpenSSL.secure_compare(expected_sig, signature_hash)
```

Error:

```
lib/workos/webhooks.rb:96: Method secure_compare does not exist on T.class_of(OpenSSL) https://srb.help/7003
    96 |        unless OpenSSL.secure_compare(expected_sig, signature_hash)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  Got T.class_of(OpenSSL) originating from:
    lib/workos/webhooks.rb:96:
    96 |        unless OpenSSL.secure_compare(expected_sig, signature_hash)
                       ^^^^^^^
Errors: 1
```

Reply:
So looks like Sorbet's typings for `OpenSSL` don't include `secure_compare`. But actually, that's a good reminder that older versions of Ruby won't have this method in their standard library (looks to be added in 2.4). So maybe we can implement our own to keep that backwards compatibility.

Reply:
I used the ActiveSupport method as a guide, but had to insert a type check to pass sorbet tests:

https://github.com/rails/rails/blob/fe76a95b0d252a2d7c25e69498b720c96b243ea2/activesupport/lib/active_support/security_utils.rb#L26

Reply:
> but had to insert a type check to pass sorbet tests

Ah, yeah. They type `String#unpack` as `T.nilable(T.any(Integer, Float, String))`, which is like the union of all the the various types that the `unpack` directives could returned. However, since the code uses `C[Int]`, I think it'll always be an array of integers.

I think the implementation from ActiveSupport is probably already correct, given its wide usage, even if Sorbet has trouble validating it. I'm also not sure if it's worth trying to avoid the overhead of the `is_a?` type-check for each byte in the string.

For example, we could do this:

```rb
      def secure_compare(str_a:, str_b:)
        return false unless str_a.bytesize == str_b.bytesize

        l = T.cast(str_a.unpack("C#{str_a.bytesize}"), T::Array[Integer])

        res = 0
        str_b.each_byte do |byte|
          v = T.must(l.shift)

          res |= byte ^ v
        end

        res.zero?
      end
```

Or, if we want to keep the implementation even closer to the original, just opt-out with `T.unsafe`:

```rb
      def secure_compare(str_a:, str_b:)
        return false unless str_a.bytesize == str_b.bytesize

        l = T.unsafe(str_a.unpack "C#{str_a.bytesize}")

        res = 0
        str_b.each_byte { |byte| res |= byte ^ l.shift }

        res.zero?
      end
```

I don't feel strongly either way. I just wanted to explore what other options Sorbet gave us in order to avoid modifying the original algorithm.

Reply:
Thanks! I went with the second option - I like sticking as close as possible to the original algorithm 👍

Reply:
Cool! Looks good!